cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -0,0 +1,474 @@
+  const std::shared_ptr<Array>& left_ = ArrayFromJSON(type, left);
+  const std::shared_ptr<Array>& right_ = ArrayFromJSON(type, right);
+  const std::shared_ptr<Array>& expected_ = ArrayFromJSON(type, expected);
+  CheckIfElseOutput(cond, left_, right_, expected_);
+}
+
+void CheckIfElseOutputSAS(const std::shared_ptr<DataType>& type,
+                          const std::shared_ptr<Scalar>& cond, const std::string& left,
+                          const std::shared_ptr<Scalar>& right,
+                          const std::string& expected) {
+  const std::shared_ptr<Array>& left_ = ArrayFromJSON(type, left);
+  const std::shared_ptr<Array>& expected_ = ArrayFromJSON(type, expected);
+  CheckIfElseOutput(cond, left_, right, expected_);
+}
+
+void CheckIfElseOutputSSA(const std::shared_ptr<DataType>& type,
+                          const std::shared_ptr<Scalar>& cond,
+                          const std::shared_ptr<Scalar>& left, const std::string& right,
+                          const std::string& expected) {
+  const std::shared_ptr<Array>& right_ = ArrayFromJSON(type, right);
+  const std::shared_ptr<Array>& expected_ = ArrayFromJSON(type, expected);
+  CheckIfElseOutput(cond, left, right_, expected_);
+}
+
+class TestIfElseKernel : public ::testing::Test {};
+
+template <typename Type>
+class TestIfElsePrimitive : public ::testing::Test {};
+
+using PrimitiveTypes = ::testing::Types<Int8Type, UInt8Type, Int16Type, UInt16Type,
+                                        Int32Type, UInt32Type, Int64Type, UInt64Type,
+                                        FloatType, DoubleType, Date32Type, Date64Type>;
+
+TYPED_TEST_SUITE(TestIfElsePrimitive, PrimitiveTypes);
+
+TYPED_TEST(TestIfElsePrimitive, IfElseFixedSizeRand) {
+  using ArrayType = typename TypeTraits<TypeParam>::ArrayType;
+  auto type = TypeTraits<TypeParam>::type_singleton();
+
+  random::RandomArrayGenerator rand(/*seed=*/0);
+  int64_t len = 1000;
+  auto cond = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(boolean(), len, /*null_probability=*/0.01));
+  auto left = std::static_pointer_cast<ArrayType>(
+      rand.ArrayOf(type, len, /*null_probability=*/0.01));
+  auto right = std::static_pointer_cast<ArrayType>(
+      rand.ArrayOf(type, len, /*null_probability=*/0.01));
+
+  typename TypeTraits<TypeParam>::BuilderType builder;
+
+  for (int64_t i = 0; i < len; ++i) {
+    if (!cond->IsValid(i) || (cond->Value(i) && !left->IsValid(i)) ||
+        (!cond->Value(i) && !right->IsValid(i))) {
+      ASSERT_OK(builder.AppendNull());
+      continue;
+    }
+
+    if (cond->Value(i)) {
+      ASSERT_OK(builder.Append(left->Value(i)));
+    } else {
+      ASSERT_OK(builder.Append(right->Value(i)));
+    }
+  }
+  ASSERT_OK_AND_ASSIGN(auto expected_data, builder.Finish());
+
+  CheckIfElseOutput(cond, left, right, expected_data);
+}
+
+#define IF_ELSE_TEST_GEN(type, l0, l1, l2, l3, r0, r1, r2, r3, valid, valid1)            \
```

Comment:
Given you have so many cases, I think a better approach might be to do this programatically - make a `CheckIfElse` that takes in Datum arguments, and have it compare the result against a 'naive' implementation. Then loop and construct some small scalars/arrays to check all the possible combinations. You can combine that with a few handwritten tests for cases like empty arrays.

Reply:
I'd break this into two stages: first use CheckScalar should suffice to verify the (array array array) and (scalar scalar scalar) cases. The remaining cases can compare the results with manually broadcast scalars

Reply:
I will try the `CheckScalar` method (haven't had seen this so far).

Reply:
I rewrote the test cases in a more generic way. Could you please take a look at it?

Reply:
`GenerateExpected` represents a reiteration of the function's logic. I'd prefer to see baseline assertions of correct behavior in the (array array array) case using CheckScalar() with explicit expected values; I think it'd be easier to read assertions like:

```c++
CheckScalar("if_else", {boolean(), int32(), int32()}, {"[true, false, null]", "[1, 2, 3]", "[4, 5, 6]"}, int32(), "[1, 5, null]");
```

Checking various permutations of input shapes can then be separate assertions or even cases.

Reply:
okay. So, IIUC you mean, we'd take the Array-Array-Array and Scalar-Scalar-Scalar case out from the `GenerateExpected` and have it separately, isn't it?

Reply:
No, I mean you can generate the other cases given correct array-array-array using `Array::GetScalar` and `MakeArrayFromScalar`

```c++
std::shared_ptr<Array> cond = ..., left = ..., right = ..., expected = ...;
// Explicit, hand written assertion for array-array-array and scalar-scalar-scalar cases
CheckScalarUnary("if_else", {cond, left, right}, expected);

// Generate Array-Scalar-Array case
for (int64_t i = 0; i < left->length(); ++i) {
  ASSERT_OK_AND_ASSIGN(auto left_scalar, left->GetScalar(i));
  ASSERT_OK_AND_ASSIGN(auto left_array, MakeArrayFromScalar(*left_scalar, left->length()));
  ASSERT_OK_AND_ASSIGN(auto expected, IfElse(cond, left_array, right)); // array-array-array is known to be correct
  ASSERT_OK_AND_ASSIGN(auto actual, IfElse(cond, left_scalar, right));
  AssertDatumsEqual(expected, actual);
}
```